File: articles/differential-equations-population-physiological-modeling/r/chemostat.R
# Chemostat biomass-substrate model.

monod_growth <- function(substrate, mu_max, K_s) {
  mu_max * substrate / (K_s + substrate)
}

simulate_chemostat <- function(X0, S0, S_in, D, Y, mu_max, K_s, dt = 0.01, t_end = 80) {
  time <- seq(0, t_end, by = dt)
  X <- numeric(length(time))
  S <- numeric(length(time))

  X[1] <- X0
  S[1] <- S0

  for (i in 2:length(time)) {
    mu <- monod_growth(S[i - 1], mu_max, K_s)

    dX <- mu * X[i - 1] - D * X[i - 1]
    dS <- D * (S_in - S[i - 1]) - (1 / Y) * mu * X[i - 1]

    X[i] <- max(X[i - 1] + dX * dt, 0)
    S[i] <- max(S[i - 1] + dS * dt, 0)
  }

  data.frame(time = time, biomass = X, substrate = S)
}

trajectory <- simulate_chemostat(0.1, 10, 20, 0.20, 0.50, 0.80, 2.0)

summary_df <- data.frame(
  final_biomass = tail(trajectory$biomass, 1),
  final_substrate = tail(trajectory$substrate, 1),
  max_biomass = max(trajectory$biomass),
  min_substrate = min(trajectory$substrate)
)

print(round(summary_df, 5))

File: articles/differential-equations-population-physiological-modeling/python/chemostat.py
"""
Chemostat dynamics scenarios.

Run:
    python python/chemostat.py
"""

from pathlib import Path

import pandas as pd

from differential_equations_core import simulate_chemostat


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "chemostat_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)
    rows = []

    for _, row in scenarios.iterrows():
        trajectory = simulate_chemostat(
            X0=row["X0"],
            S0=row["S0"],
            S_in=row["S_in"],
            D=row["D"],
            Y=row["Y"],
            mu_max=row["mu_max"],
            K_s=row["K_s"],
            dt=row["dt"],
            t_end=row["t_end"],
        )

        rows.append(
            {
                "scenario": row["scenario"],
                "final_biomass": trajectory["biomass"].iloc[-1],
                "final_substrate": trajectory["substrate"].iloc[-1],
                "max_biomass": trajectory["biomass"].max(),
                "min_substrate": trajectory["substrate"].min(),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
